pkg/operator/ceph/cluster/osd/osd_on_pvc_test.go

```diff
@@ -0,0 +1,720 @@
+	exectest "github.com/rook/rook/pkg/util/exec/test"
+	"github.com/stretchr/testify/assert"
+	"github.com/tevino/abool"
+	appsv1 "k8s.io/api/apps/v1"
+	batchv1 "k8s.io/api/batch/v1"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/resource"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/version"
+	"k8s.io/apimachinery/pkg/watch"
+	fakediscovery "k8s.io/client-go/discovery/fake"
+	"k8s.io/client-go/kubernetes/fake"
+	k8stesting "k8s.io/client-go/testing"
+)
+
+// We want to keep track of the order in which some resources (notably deployments) are created and
+// updated, and the tracker should be thread safe since there could be operations occurring in
+// parallel.
+
+type resourceOperation struct {
+	resourceName string
+	operation    string // e.g., "create", "update"
+}
+
+func newResourceOperation(resourceName, operation string) resourceOperation {
+	return resourceOperation{resourceName, operation}
+}
+
+type resourceOperationList struct {
+	m           sync.Mutex
+	resourceOps []resourceOperation
+}
+
+func newResourceOperationList() *resourceOperationList {
+	return &resourceOperationList{
+		sync.Mutex{},
+		[]resourceOperation{},
+	}
+}
+
+func (r *resourceOperationList) Add(resourceName, operation string) {
+	r.m.Lock()
+	defer r.m.Unlock()
+	r.resourceOps = append(r.resourceOps, newResourceOperation(resourceName, operation))
+}
+
+func (r *resourceOperationList) Len() int {
+	r.m.Lock()
+	defer r.m.Unlock()
+	return len(r.resourceOps)
+}
+
+func (r *resourceOperationList) List() []resourceOperation {
+	return r.resourceOps
+}
+
+// just list the resources operated on
+func (r *resourceOperationList) ResourcesWithOperation(operation string) []string {
+	resources := []string{}
+	for _, ro := range r.List() {
+		if ro.operation == operation {
+			resources = append(resources, ro.resourceName)
+		}
+	}
+	return resources
+}
+
+func (r *resourceOperationList) OperationsInOrder() []string {
+	ops := []string{}
+	for _, ro := range r.List() {
+		ops = append(ops, ro.operation)
+	}
+	return ops
+}
+
+// The definition for this test is a wrapper for the test function that adds a timeout
+func TestOSDsOnPVC(t *testing.T) {
+	// This test runs in less than 350 milliseconds on a 6-core CPU w/ 16GB RAM
+	done := make(chan bool)
+	timeout := time.After(5 * 350 * time.Millisecond)
+
+	go func() {
+		// use defer because t.Fatal will kill this goroutine, and we always want done set if the
+		// func stops running
+		defer func() { done <- true }()
+		// run the actual test
+		testOSDsOnPVC(t)
+	}()
+
+	select {
+	case <-timeout:
+		t.Fatal("Test timed out. This is a test failure.")
+	case <-done:
+	}
+}
+
+// This is the actual test. If it hangs, we should consider that an error. Writing a timeout in
+// tests requires running the test in a goroutine, so there is a timeout wrapper above
+func testOSDsOnPVC(t *testing.T) {
```

Comment:
Wow, a 600 line func. :) How can we break this down for readability? For example, there are many inline functions that could be separated, including all the reactors.

Reply:
Yeah... I agree.

I have had a thought that might help. How do you feel about...
For the Jobs creating/deleting Pods and the PVCs creating a name from `generateName` I was thinking that we could make a more complex kube clientset for unit tests that is reusable between tests where we might need the test clientset to do some things that are closer to how k8s does things in the real world.

Reply:
Yes that could certainly be helpful to have a client the creates the resources with a predictable generateName. In #6982 I made it an incrementing index for predictability in the test instead of generating a random name, not sure if that approach would be useful to re-use for this test. If not, I'm sure there is some other commonality that we could find across tests.

Reply:
I was also able to add functionality that creates a pod for each job.